File: scripts/canteen.mjs
// 福建师范大学食堂空座率抓取（Node 版，crawler/canteen.py 的回退实现）
// 尽力抓取福star「食堂人流量分析」接口（校园网环境），输出 public/data/canteen_live.json。
// 失败不覆盖旧数据（尽力而为），以非零码退出。
// 运行：node scripts/canteen.mjs
import fs from 'node:fs'
import path from 'node:path'
import { fileURLToPath } from 'node:url'

const __dirname = path.dirname(fileURLToPath(import.meta.url))
const OUT = path.join(__dirname, '..', 'public', 'data', 'canteen_live.json')
const UA =
  'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0 Safari/537.36'
const CANDIDATE_URLS = [
  'https://canteen.fjnu.edu.cn/api/live',
  'https://fstar.fjnu.edu.cn/api/canteen/live'
]

function normalize(rawItems) {
  const out = []
  for (const it of rawItems || []) {
    if (!it || typeof it !== 'object') continue
    const name = it.name || it.canteen || it.title || ''
    const people = it.people ?? it.count ?? it.current ?? 0
    const seats = it.seats ?? it.capacity ?? 0
    const daily = it.daily ?? it.consumes ?? it.consumption ?? 0
    if (!name) continue
    out.push({ name, people: Number(people) || 0, seats: Number(seats) || 0, daily: Number(daily) || 0 })
  }
  return out
}

async function main() {
  let lastErr = null
  for (const url of CANDIDATE_URLS) {
    try {
      const ctrl = new AbortController()
      const timer = setTimeout(() => ctrl.abort(), 10000)
      const res = await fetch(url, { signal: ctrl.signal, headers: { 'User-Agent': UA } })
      clearTimeout(timer)
      if (!res.ok) throw new Error('HTTP ' + res.status)
      const d = await res.json()
      const items = normalize(Array.isArray(d) ? d : d.items)
      if (!items.length) throw new Error('empty items')
      const snap = {
        status: 'live',
        updatedAt: new Date().toISOString(),
        items
      }
      fs.mkdirSync(path.dirname(OUT), { recursive: true })
      fs.writeFileSync(OUT, JSON.stringify(snap))
      console.log(`canteen live written: ${OUT} (${items.length} 家)`)
      return
    } catch (e) {
      lastErr = e
    }
  }
  if (fs.existsSync(OUT)) console.log('canteen live unavailable（校园网环境），保留上一次快照')
  else console.log('canteen live unavailable（校园网环境），无旧快照')
  console.log('last error:', lastErr ? lastErr.message : 'n/a')
  process.exit(1)
}

main()

File: crawler/canteen.py
# -*- coding: utf-8 -*-
"""FJNU 校园导航 · 食堂空座率抓取（Python 版）

尽力抓取福建师范大学「食堂人流量分析」接口（福Star APP 公共服务，需校园网环境），
生成 `public/data/canteen_live.json`（与前端 canteens.js 的 name 匹配）。

校外/无校园网环境抓取失败时**不覆盖**上一次成功数据，并输出非零退出码供 CI 判断
（该步骤在 snapshot.yml 中 continue-on-error，不阻塞整体）。

用法：
    python crawler/canteen.py
"""
import json
import sys
from datetime import datetime, timezone
from pathlib import Path

sys.stdout.reconfigure(encoding='utf-8')
ROOT = Path(__file__).resolve().parent.parent
sys.path.insert(0, str(ROOT))

OUT_FILE = ROOT / 'public' / 'data' / 'canteen_live.json'

# 福star食堂人流量分析接口（校园网环境）。形如福Star 服务域名，字段按
# { name, people, seats, daily } 输出，与前端匹配。
CANDIDATE_URLS = [
    'https://canteen.fjnu.edu.cn/api/live',
    'https://fstar.fjnu.edu.cn/api/canteen/live',
]

try:
    import requests
except Exception:  # noqa: BLE001
    requests = None

UA = ('Mozilla/5.0 (Windows NT 10.0; Win64; x64) '
      'AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0 Safari/537.36')


def _get(url, timeout=10):
    if requests is not None:
        r = requests.get(url, headers={'User-Agent': UA}, timeout=timeout, verify=True)
        r.raise_for_status()
        return r.json()
    import urllib.request
    req = urllib.request.Request(url, headers={'User-Agent': UA})
    with urllib.request.urlopen(req, timeout=timeout) as resp:
        return json.loads(resp.read().decode('utf-8'))


def normalize(raw_items):
    """归一化接口返回 → [{name, people, seats, daily}]（容忍字段名差异）。"""
    out = []
    for it in raw_items or []:
        if not isinstance(it, dict):
            continue
        name = it.get('name') or it.get('canteen') or it.get('title') or ''
        people = it.get('people') or it.get('count') or it.get('current') or 0
        seats = it.get('seats') or it.get('capacity') or 0
        daily = it.get('daily') or it.get('consumes') or it.get('consumption') or 0
        if not name:
            continue
        out.append({'name': name, 'people': int(people or 0),
                    'seats': int(seats or 0), 'daily': int(daily or 0)})
    return out


def main():
    last_err = None
    for url in CANDIDATE_URLS:
        try:
            d = _get(url)
            items = normalize(d.get('items') if isinstance(d, dict) else d)
            if not items:
                raise ValueError('empty items')
            snap = {
                'status': 'live',
                'updatedAt': datetime.now(timezone.utc).isoformat().replace('+00:00', 'Z'),
                'items': items,
            }
            OUT_FILE.parent.mkdir(parents=True, exist_ok=True)
            OUT_FILE.write_text(json.dumps(snap, ensure_ascii=False, separators=(',', ':')), encoding='utf-8')
            print('canteen live written: %s (%d 家)' % (OUT_FILE, len(items)))
            return
        except Exception as exc:  # noqa: BLE001 单个候选失败继续
            last_err = exc
            continue
    # 全部失败：保留旧文件（若有），非零退出
    if OUT_FILE.exists():
        print('canteen live unavailable（校园网环境），保留上一次快照')
    else:
        print('canteen live unavailable（校园网环境），无旧快照')
    print('last error: %s' % (last_err or 'n/a'))
    sys.exit(1)


if __name__ == '__main__':
    main()